optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/chained/ChainedSwapMove.java

```diff
@@ -74,8 +77,8 @@ protected void doMoveOnGenuineVariables(ScoreDirector<Solution_> scoreDirector)
                     scoreDirector.changeVariableFacade(variableDescriptor, leftEntity, oldRightValue);
                     scoreDirector.changeVariableFacade(variableDescriptor, rightEntity, oldLeftValue);
                 } else {
-                    Object oldLeftTrailingEntity = oldLeftTrailingEntityList.get(i);
-                    Object oldRightTrailingEntity = oldRightTrailingEntityList.get(i);
+                    Object oldLeftTrailingEntity = oldLeftTrailingEntityMap.get(variableDescriptor);
```

Comment:
These map calls are a performance loss. **We're replacing an indexed lookup by a hash lookup.**
I'd keep the index lookup (so the trailingEntityList has the same size as variableDescriptorList and their indexes match), but given that "Object oldLeftTrailingEntity " can be null, I 'd put a null as an element where we cache that (move constructor)